taca/illumina/HiSeqX_Runs.py

```diff
@@ -63,102 +66,111 @@ def _copy_samplesheet(self):
 
     def demultiplex_run(self):
         """
-           Demultiplex a Xten run:
-            - find the samplesheet
-            - make a local copy of the samplesheet and name it SampleSheet.csv
-            - define if necessary the bcl2fastq commands (if indexes are not of size 8, i.e. neoprep)
+           Demultiplex a run:
+            - Make sub-samplesheet based on sample classes
+            - Decide correct bcl2fastq command parameters based on sample classes
             - run bcl2fastq conversion
         """
-        #we have 10x lane - need to split the  samples sheet and build a 10x command for bcl2fastq
-        Complex_run = False
-        if len(self.lanes_10X) and len(self.lanes_not_10X):
-             Complex_run = True
-
-        if Complex_run:
-            with chdir(self.run_dir):
-                samplesheet_dest_not_10X="SampleSheet_0.csv"
-                with open(samplesheet_dest_not_10X, 'wb') as fcd:
-                    fcd.write(_generate_samplesheet_subset(self.runParserObj.samplesheet, self.lanes_not_10X, self.tenX_samples, is_10X = False))
-                samplesheet_dest_10X="SampleSheet_1.csv"
-                with open(samplesheet_dest_10X, 'wb') as fcd:
-                    fcd.write(_generate_samplesheet_subset(self.runParserObj.samplesheet, self.lanes_10X, self.tenX_samples, is_10X = True))
-        else:
-            with chdir(self.run_dir):
-                samplesheet_dest="SampleSheet_0.csv"
-                with open(samplesheet_dest, 'wb') as fcd:
-                    if len(self.lanes_10X):
-                        fcd.write(_generate_samplesheet_subset(self.runParserObj.samplesheet, self.lanes_10X, self.tenX_samples, is_10X = True))
-                    else:
-                        fcd.write(_generate_samplesheet_subset(self.runParserObj.samplesheet, self.lanes_not_10X, self.tenX_samples, is_10X = False))
-
-        per_lane_base_masks = self._generate_per_lane_base_mask()
-        max_different_base_masks =  max([len(per_lane_base_masks[base_masks]) for base_masks in per_lane_base_masks])
-        if max_different_base_masks > 1:
-            # in a HiSeqX run I cannot have different index sizes in the SAME lane
-            logger.error("In FC {} found one or more lane with more than one base mask (i.e., different index sizes in \
-                         in the same lane".format(self.id))
-            return False
+        # Check sample types
+        sample_type_list = []
+        for lane, lane_contents in self.sample_table.items():
+            for sample in lane_contents:
+                if sample[sample.keys()[0]]['type'] not in sample_type_list:
```

Comment:
Just a thought - will `sample.keys()[0]` always return the desired value? Isn't the order of dicts random in python 2.7?

Reply:
Hi, yes. sample.keys() is a list with one single value

Reply:
Ah, I see now.